presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -755,8 +755,9 @@ public boolean transitionToFinishing()
             return false;
         }
 
-        if (autoCommit) {
-            ListenableFuture<?> commitFuture = transactionManager.asyncCommit(session.getTransactionId().get());
+        Optional<TransactionId> transactionId = session.getTransactionId();
+        if (transactionId.isPresent() && transactionManager.transactionExists(transactionId.get()) && transactionManager.isAutoCommit(transactionId.get())) {
```

Comment:
I thought there might be a race between the two calls, but it seems the only way to remove the transaction is to call `asyncAbort` which is guarded by the state machine.

Reply:
the extra call is required for `ROLLBACK` which immediately removes the transaction from the manager, so without the exists check the call to `isAutoCommit` fails.  I looked as having an explicit check for `ROLLBACK` here, but that would require breaking to many abstractions.  I also looked at making `isAutoCommit`, return Optional, or just false, for unknown queries, but that makes is difficult to work with in the normal case.